src/osd/PrimaryLogPG.cc

```diff
@@ -1529,6 +1529,46 @@ int PrimaryLogPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
   return r;
 }
 
+// it is pretty much the same as CEPH_OSD_OP_COPY_FROM, but it requires user to
+// specify the src osd, and uses repair_read() to read the object payload, also
+// the src oid is always identical to the dest oid (the one in OpContext)
+int PrimaryLogPG::do_repair_copy(OpContext *ctx, const OSDOp& osd_op, bufferlist::iterator& bp)
+{
+  repair_copy_arg_t arg;
+
+  vector<pg_shard_t> bad_shards;
+  try {
+    arg.decode(bp);
+  } catch (buffer::error&) {
+    dout(10) << "corrupted " << __func__ << " arg" << dendl;
+    return -EINVAL;
+  }
+  if (arg.bad_shards.empty() || !arg.what) {
+    return 0;
+  }
+  const auto& oi = ctx->obs->oi;
+  if (!ctx->copy_cb) {
+    // mark the bad shard missing so they are not chosen when repairing
+    if (pool.info.require_rollback()) {
```

Comment:
I think here you really mean pool.info.ec_pool().  Does it actually matter though?  Why not add it to the peer_missing sets unconditionally so that Replicated and EC work uniformly?